From 013ad5327960c6b18694199cd85a56bb8b6eb4f4 Mon Sep 17 00:00:00 2001
From: Peter Hutterer <peter.hutterer@who-t.net>
Date: Thu, 30 Jun 2011 14:41:52 +1000
Subject: [PATCH evdev 2/4] Revert "Disable middle mouse button emulation by
 default."

RHEL6.0 shipped with MB emulation enabled (auto), don't change this
behaviour.

This reverts commit 21a2ac818e75ef918d320ce1e88b6263e68e598d.
---
 man/evdev.man |  3 ++-
 src/emuMB.c   | 37 ++++++++++++++++++++++++++++++++++---
 src/evdev.h   |  1 +
 3 files changed, 37 insertions(+), 4 deletions(-)

diff --git a/man/evdev.man b/man/evdev.man
index 220dd13..fa08406 100644
--- a/man/evdev.man
+++ b/man/evdev.man
@@ -86,7 +86,8 @@ indicating that the next button pressed is to be
 .BI "Option \*qEmulate3Buttons\*q \*q" boolean \*q
 Enable/disable the emulation of the third (middle) mouse button for mice
 which only have two physical buttons.  The third button is emulated by
-pressing both buttons simultaneously.  Default: off. Property: "Evdev Middle
+pressing both buttons simultaneously.  Default: off for touchscreens, otherwise
+on until a middle mouse button event is registered. Property: "Evdev Middle
 Button Emulation".
 .TP 7
 .BI "Option \*qEmulate3Timeout\*q \*q" integer \*q
diff --git a/src/emuMB.c b/src/emuMB.c
index b25eac8..623561b 100644
--- a/src/emuMB.c
+++ b/src/emuMB.c
@@ -43,6 +43,12 @@
 
 #include <evdev-properties.h>
 
+enum {
+    MBEMU_DISABLED = 0,
+    MBEMU_ENABLED,
+    MBEMU_AUTO
+};
+
 static Atom prop_mbemu     = 0; /* Middle button emulation on/off property */
 static Atom prop_mbtimeout = 0; /* Middle button timeout property */
 /*
@@ -225,6 +231,11 @@ EvdevMBEmuFilterEvent(InputInfoPtr pInfo, int button, BOOL press)
     if (!pEvdev->emulateMB.enabled)
         return ret;
 
+    if (button == 2) {
+        EvdevMBEmuEnable(pInfo, FALSE);
+        return ret;
+    }
+
     /* don't care about other buttons */
     if (button != 1 && button != 3)
         return ret;
@@ -299,9 +310,20 @@ EvdevMBEmuPreInit(InputInfoPtr pInfo)
 {
     EvdevPtr pEvdev = (EvdevPtr)pInfo->private;
 
-    pEvdev->emulateMB.enabled = xf86SetBoolOption(pInfo->options,
-                                                  "Emulate3Buttons",
-                                                  FALSE);
+    if (pEvdev->flags & EVDEV_TOUCHSCREEN)
+        pEvdev->emulateMB.enabled = MBEMU_DISABLED;
+    else
+        pEvdev->emulateMB.enabled = MBEMU_AUTO;
+
+    if (xf86FindOption(pInfo->options, "Emulate3Buttons"))
+    {
+        pEvdev->emulateMB.enabled = xf86SetBoolOption(pInfo->options,
+                                                      "Emulate3Buttons",
+                                                      MBEMU_ENABLED);
+        xf86Msg(X_INFO, "%s: Forcing middle mouse button emulation %s.\n",
+                pInfo->name, (pEvdev->emulateMB.enabled) ? "on" : "off");
+    }
+
     pEvdev->emulateMB.timeout = xf86SetIntOption(pInfo->options,
                                                  "Emulate3Timeout", 50);
 }
@@ -329,6 +351,15 @@ EvdevMBEmuFinalize(InputInfoPtr pInfo)
 
 }
 
+/* Enable/disable middle mouse button emulation. */
+void
+EvdevMBEmuEnable(InputInfoPtr pInfo, BOOL enable)
+{
+    EvdevPtr pEvdev = (EvdevPtr)pInfo->private;
+    if (pEvdev->emulateMB.enabled == MBEMU_AUTO)
+        pEvdev->emulateMB.enabled = enable;
+}
+
 static int
 EvdevMBEmuSetProperty(DeviceIntPtr dev, Atom atom, XIPropertyValuePtr val,
                       BOOL checkonly)
diff --git a/src/evdev.h b/src/evdev.h
index 4742b43..cc479dc 100644
--- a/src/evdev.h
+++ b/src/evdev.h
@@ -277,6 +277,7 @@ void EvdevMBEmuBlockHandler(pointer, struct timeval**, pointer);
 void EvdevMBEmuPreInit(InputInfoPtr);
 void EvdevMBEmuOn(InputInfoPtr);
 void EvdevMBEmuFinalize(InputInfoPtr);
+void EvdevMBEmuEnable(InputInfoPtr, BOOL);
 
 /* Third button emulation */
 CARD32 Evdev3BEmuTimer(OsTimerPtr timer, CARD32 time, pointer arg);
-- 
1.9.0

